feat: complifi strategy

fp-crypto opened the pull request:
The `ComplifiUSDC` strategy uses USDC to mint equal numbers of fully collateralized 5x leveraged tokens against an underlying (e.g. 2 usdc = 1 5x eth up and 1 5x eth down token). These tokens can be refunded at any time for the same amount of collateral, assuming they are provided in equal numbers. Holding equal numbers of up and down tokens carries no directional risk. Complifi is offering farm rewards in their protocol token, COMFI. The strategy: 

 - Mints 5x leveraged up and down tokens with USDC
 - Stakes them for COMFI rewards
 - Sells the rewards for more USDC
 - Redeems the tokens on demand, by:
     - Withdrawing from the rewards contract
     - Using the vault refund method to redeem equal amounts of up and down tokens for USDC
     
## Notes

We are targeting `yearn-vaults@0.3.0` as USDC prod is on that version.

Commit:
feat: initial work

Commit:
test: strat

Commit:
fix: lots

Commit:
feat: working?

Commit:
feat: works

Commit:
chore: cleanup and comment

Commit:
fix: less mining

Commit:
test: fix

Commit:
chore: cleanup

Commit:
test: moar tests

Commit:
feat: add migrateTokenVault functionality

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -16,32 +16,106 @@ import {
+                primaryToken().balanceOf(address(this))
+            );
+        (uint256 depositedPrimary, ) =
+            liquidityMining.userPoolInfo(primaryTokenPid(), address(this));
```
you can use `liquidityMining.userPoolInfo(primaryTokenPid(), address(this)).amount;`

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -16,32 +16,106 @@ import {
-        // profitFactor = 100;
-        // debtThreshold = 0;
+        want.safeApprove(address(tokenVault), type(uint256).max);
+        primaryToken().safeApprove(address(tokenVault), type(uint256).max);
```
when you use a method more than once, assign to a local var.

fp-crypto replied:
do you mean `address(tokenVault)`?

poolpitako replied:
```
address primaryToken = primaryToken();

primaryToken.firstUsage();
primaryToken.secondUsage();
```

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
+        uint256 assets = estimatedTotalAssets();
+        uint256 wantBal = want.balanceOf(address(this));
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
```
we need to be careful with this when downgrading to 0.3.0

fp-crypto replied:
what changes?

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
 
-    // NOTE: Can override `tendTrigger` and `harvestTrigger` if necessary
+    function migrateTokenVault(address _newTokenVault) external onlyGovernance {
+        require(
```
I think you can kill this. we wouldn't migrate something to a wrong state.

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
+            IComplifiVault(_newTokenVault).state() == IComplifiVault.State.Live
+        );
+
+        liquidatePosition(uint256(-1));
```
`type(uint256).max`

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
-    //      protected[2] = tokenC;
-    //      return protected;
-    //    }
+        liquidatePosition(uint256(-1)); //withdraw all. does not matter if we ask for too much
```
`type(uint256).max`

poolpitako commented on contracts/Strategy.sol:
```diff
@@ -53,54 +127,213 @@ contract Strategy is BaseStrategy {
+        liquidatePosition(uint256(-1));
+
+        // Revoke approvals for the old token vault
+        want.safeApprove(address(tokenVault), type(uint256).min);
```
`0` instead of `type(uint256).min`?

fp-crypto replied:
Moved approvals to a helper because DRY

Commit:
fix: address @poolpitako comments

Commit:
feat: rollback to vault 0.3.0

Commit:
feat: parameterized initializer

Commit:
feat: clone strategy

Commit:
test: more coverage of clone

Commit:
fix: @dudesahn's comments

https://github.com/yearn/yearn-strategies/issues/70#issuecomment-840804889
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-840808701

Commit:
feat: check if token vault is in registry

per @dudesahn's suggestion
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-840816428

Commit:
chore: improve path logic

Commit:
fix: only use safemath

@jmonteer's comment
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-841109171
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-841113850

Commit:
fix: @jmonteer's comments

https://github.com/yearn/yearn-strategies/issues/70#issuecomment-841122869
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-841144118
https://github.com/yearn/yearn-strategies/issues/70#issuecomment-841133401

Commit:
chore: test cleanup

Commit:
chore: pare down to save gas
